homeassistant/components/tag/__init__.py

```diff
@@ -105,10 +110,62 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
         storage_collection, DOMAIN, DOMAIN, CREATE_FIELDS, UPDATE_FIELDS
     ).async_setup(hass)
 
+    entities: dict[str, TagEntity] = {}
+
+    async def tag_change_listener(
+        change_type: str, item_id: str, updated_config: dict
+    ) -> None:
+        """Tag event listener."""
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            _LOGGER.debug(
+                "%s, item: %s, update: %s", change_type, item_id, updated_config
+            )
+        if change_type == collection.CHANGE_ADDED:
+            # When tags are added to storage
+            entities[updated_config[TAG_ID]] = TagEntity(
+                hass,
+                updated_config.get(CONF_NAME, DEFAULT_NAME),
+                updated_config[TAG_ID],
+                updated_config.get(LAST_SCANNED, ""),
+                updated_config.get(DEVICE_ID),
+            )
+            await entities[updated_config[TAG_ID]].async_add_initial_state()
+
+        if change_type == collection.CHANGE_UPDATED:
+            # When tags are changed or updated in storage
+            if entities[updated_config[TAG_ID]]._last_scanned != updated_config.get(  # pylint: disable=protected-access
+                LAST_SCANNED, ""
+            ):
+                entities[updated_config[TAG_ID]].async_handle_event(
+                    updated_config.get(DEVICE_ID),
+                    updated_config.get(LAST_SCANNED, ""),
+                )
+
+        # Deleted tags
+        if change_type == collection.CHANGE_REMOVED:
+            # When tags is removed from storage
+            await entities[updated_config[TAG_ID]].async_remove()
```

Comment:
This sets the entity to `unavailable` because the entity is in the entity registry, then, just below, the entity is removed - twice - from the state machine.

This all seems odd, what do we actually want to happen when the tag is removed? With the current code, the entity will be back after a restart of Home Assistant, with the state set to `unavailable` because the entity is in the entity registry.

If we want the state to be removed, we should remove the entity from the entity registry instead of removing it from the state machine.

Reply:
Some copy/paste errors from previous attempts when this was in the entity itself. Corrected shortly

Reply:
These entities are not recorded in entity registry. Do we want that?